src/brod_producer.erl

```diff
@@ -48,459 +35,84 @@
-             , ack_timeout = AckTimeout
-             , client_id   = ensure_binary(ClientId)
-             }}.
+  {ok, Metadata} = brod_client:get_metadata(ClientId, Topic),
+  #metadata_response{topics = [TopicMetadata]} = Metadata,
+  #topic_metadata{ error_code = TopicErrorCode
+                 , partitions = PartitionsMetadataList
+                 } = TopicMetadata,
+  brod_kakfa:is_error(TopicErrorCode) orelse
+    erlang:throw({"topic metadata error", TopicErrorCode}),
+  Partitions = lists:map(fun(#partition_metadata{id = Id}) -> Id end,
+                         PartitionsMetadataList),
+  PartitionProducers =
+    [start_partition_producer(ClientId, Topic, P, Config) || P <- Partitions],
+  #state{ client_id  = ClientId
+        , topic      = Topic
+        , config     = Config
+        , partitions = PartitionProducers
+        }.
 
 handle_call(stop, _From, State) ->
   {stop, normal, ok, State};
-handle_call(get_sockets, _From, State) ->
-  {reply, {ok, State#state.sockets}, State};
-handle_call({produce, Sender, Ref, Topic, Partition, KVList}, _From, State0) ->
-  case handle_produce(Sender, Ref, Topic, Partition, KVList, State0) of
-    {ok, Leader, State1} ->
-      {ok, State} = maybe_send(Leader, State1),
-      {reply, {ok, Ref}, State};
-    {error, Error} ->
-      {reply, {error, Error}, State0}
-  end;
 handle_call(Request, _From, State) ->
   {reply, {error, {unsupported_call, Request}}, State}.
 
-handle_cast(_Msg, State) ->
+handle_cast({produce, _Caller, _Key, _Value}, State) ->
+  %% TODO round-robin, random, etc.
+  {noreply, State};
+handle_cast(_Cast, State) ->
   {noreply, State}.
 
-handle_info({msg, Leader, CorrId, #produce_response{} = R}, State0) ->
-  Errors = get_error_codes(R#produce_response.topics),
-  case handle_produce_response(Leader, CorrId, Errors, State0) of
-    {ok, State} ->
-      {noreply, State};
-    {error, Reason} ->
-      %% Errors found in produce response, terminate myself.
-      %% In a nicer way, it should recognize at least error code
-      %% ?EC_NOT_LEADER_FOR_PARTITION, then try to connect to the
-      %% new leader (if not yet connected) and migrate data buffer
-      %% and pending list from the old leader to the new leader.
-      %% We can not do this now because the payload is not kept in
-      %% the pending list for resend, there is no better solution
-      %% than crash myself and hope the brod clients would (and they
-      %% should) retry after restart.
-      {stop, Reason, State0}
-  end;
-handle_info({'EXIT', Pid, Reason}, State) ->
-  Sockets = lists:keydelete(Pid, #socket.pid, State#state.sockets),
-  %% Shutdown because cluster has elected other leaders for
-  %% every topic/partition most likely.
-  %% Users will have to resend outstanding messages
-  {stop, {socket_down, Reason}, State#state{sockets = Sockets}};
-handle_info(_Info, State) ->
+handle_info({'EXIT', _Pid, _Reason}, #state{} = State) ->
+  %% TODO: handle partition producer restart
```

Comment:
why not start them under a simple one for one supervisor and let that one to handle restarts?